### PR TITLE
Mic 4385/port 1040 formatting

### PR DESCRIPTION
**Pull request by albrja**
### Moves 1040 dataset formatting to post-processing
- *Category*: Post-processing
- *JIRA issue*: [MIC-4385](https://jira.ihme.washington.edu/browse/MIC-4385)
- *Research reference*: 

### Changes and notes
-moves 1040 data formatting to post-processing
-updates get_itin_map to get ssn_itin_map to combine ssn and itin columns for 1040
-no longer writes tax dependents dataset to file since it is no longer needed for pseudopeople

### Verification and Testing
Successfully ran make results and saw expected columns.

### CODE REVIEW

**Review by rmudambi — APPROVED** [review]

File: src/vivarium_census_prl_synth_pop/results_processing/formatter.py [hunk 1]
```diff
@@ -162,3 +162,105 @@ def format_data_for_mapping(
...
+    # Choose 4 dependents
+    data = data.loc[data["rank"] < 5]
+    data["rank"] = data["rank"].astype(str)
+    flat = data.pivot(columns="rank", index=index_cols, values=value_cols)
+    flat.columns = ["_".join([pair[1], pair[0]]) for pair in flat.columns]
+
+    return flat
+
+
+def combine_joint_filers(data: pd.DataFrame) -> pd.DataFrame:
```
Comment (stevebachmeier) [inline src/vivarium_census_prl_synth_pop/results_processing/formatter.py:244]: I assume a followup PR will be removing all this from pseudopeople?

File: src/vivarium_census_prl_synth_pop/tools/make_results.py [hunk 2]
```diff
@@ -371,11 +379,21 @@ def perform_post_processing(
...
+        processed_results
+    )
 
-        logger.info(f"Writing final {observer} results.")
-        obs_dir = build_output_dir(final_output_dir, subdir=observer)
-        seed_ext = f"_{seed}" if seed != "" else ""
-        write_to_disk(obs_data.copy(), obs_dir / f"{observer}{seed_ext}.{extension}")
+    # Write results for each dataset - we do not need to write out tax dependents now that we format
+    # the 1040 dataset above
+    for observer in FINAL_OBSERVERS:
```
Comment (stevebachmeier) [inline src/vivarium_census_prl_synth_pop/tools/make_results.py:390]: Aren't you already in a `for observer in FINAL_OBSERVERS` loop (line 330)?

File: src/vivarium_census_prl_synth_pop/tools/make_results.py [hunk 3]
```diff
@@ -371,11 +379,21 @@ def perform_post_processing(
...
+        if observer in OUTPUT_DATASETS:
+            obs_data = processed_results[observer]
```
Comment (stevebachmeier) [inline src/vivarium_census_prl_synth_pop/tools/make_results.py:392]: You also already have obs_data loaded (though I think you did this b/c the 1040 stuff has changed since initial loading?)

File: src/vivarium_census_prl_synth_pop/tools/make_results.py [hunk 4]
```diff
@@ -371,11 +379,21 @@ def perform_post_processing(
...
+    )
 
-        logger.info(f"Writing final {observer} results.")
-        obs_dir = build_output_dir(final_output_dir, subdir=observer)
-        seed_ext = f"_{seed}" if seed != "" else ""
-        write_to_disk(obs_data.copy(), obs_dir / f"{observer}{seed_ext}.{extension}")
+    # Write results for each dataset - we do not need to write out tax dependents now that we format
+    # the 1040 dataset above
+    for observer in FINAL_OBSERVERS:
+        if observer in OUTPUT_DATASETS:
```
Comment (stevebachmeier) [inline src/vivarium_census_prl_synth_pop/tools/make_results.py:391]: Why this if statement? I think FINAL_OBSERVERS and OUTPUT_DATASETS have the same seven items.

**Review by stevebachmeier — CHANGES_REQUESTED** [review]
A couple minor suggestions/questions.

File: src/vivarium_census_prl_synth_pop/results_processing/formatter.py (hunk 1 above)
Comment (albrja) [reply src/vivarium_census_prl_synth_pop/results_processing/formatter.py:244]: Yes

File: src/vivarium_census_prl_synth_pop/tools/make_results.py (hunk 2 above)
Comment (albrja) [reply src/vivarium_census_prl_synth_pop/tools/make_results.py:390]: Yes. I had to format the 1040 and needed both the 1040 and tax dependents to do so I had to leave the loop once all of them were mapped.

File: src/vivarium_census_prl_synth_pop/tools/make_results.py (hunk 3 above)
Comment (albrja) [reply src/vivarium_census_prl_synth_pop/tools/make_results.py:392]: Yes

File: src/vivarium_census_prl_synth_pop/tools/make_results.py (hunk 4 above)
Comment (albrja) [reply src/vivarium_census_prl_synth_pop/tools/make_results.py:391]: OUTPUT_DATASETS doesn't have tax dependents since do not need to write it out anymore. It was this or special casing not writing tax dependents.

File: tests/test_data_utils.py [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+            "spouse_joint_filer",
+            "spouse_tax_year",
+            "spouse_household_id",
+            "spouse_ssn_itin",
+            "spouse_copy_ssn",
+        ]
+    )
+    joint_filer_ids = dummy_1040.loc[dummy_1040["joint_filer"] == True, "simulant_id"]
+    # Joint filer ids should not be in simulant ids
+    assert not bool(set(joint_1040["simulant_id"]) & set(joint_filer_ids))
```
Comment (stevebachmeier) [inline tests/test_data_utils.py:113]: You don't actually need to cast the thing as a bool.

Comment (stevebachmeier) [reply tests/test_data_utils.py:113]: e.g. 

>>> assert not {1,2,3} & {4,5,6}
>>> assert not {1,2,3} & {4,5,6,1}
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AssertionError
>>>

File: tests/test_data_utils.py [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+from vivarium_census_prl_synth_pop.results_processing.formatter import (
+    combine_joint_filers,
+    flatten_data,
+    format_1040_dataset,
+)
+
+
+@pytest.fixture(scope="module")
+def dummy_1040():
+    return pd.DataFrame(
```
Comment (stevebachmeier) [inline tests/test_data_utils.py:14]: consider adding docstrings to these dummy datasets since they're kinda complicated and explain why you did what you did (i.e. what edge cases you're trying to test against)

File: tests/test_data_utils.py [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+                False,
+                False,
+            ]
+            * 2,
+            "relationship_to_reference_person": [
+                "Reference person",
+                "Opp-sex spouse",
+                "Biological child",
+                "Reference person",
+                "Opp-sex spouse",
```
Comment (stevebachmeier) [inline tests/test_data_utils.py:33]: same-sex spouse?

Comment (stevebachmeier) [reply tests/test_data_utils.py:33]: or maybe even make totally generic to avoid convusion and unnecessary thinking, e.g. "relationship 1", "relationship 2"

**Review by stevebachmeier — APPROVED** [review]
